distributed/shuffle/shuffle_scheduler.py

```diff
@@ -0,0 +1,217 @@
+
+            try:
+                restrictions = {
+                    self.worker_for_key(dts.key, state.out_tasks_left, state.workers)
+                }
+            except (RuntimeError, IndexError, ValueError) as e:
+                raise type(e)(
+                    f"Could not pick worker to run dependent {dts.key} of {key}: {e}"
+                ) from None
+
+            assert (
+                not dts.worker_restrictions or dts.worker_restrictions == restrictions
+            ), (
+                f"Output task {dts.key} (of shuffle {id}) has unexpected worker restrictions "
+                f"{dts.worker_restrictions}, not {restrictions}"
+            )
+            # TODO if these checks fail, we need to error the task!
+            # Otherwise it'll still run, and maybe even succeed, but just produce wrong data?
+
+            dts._worker_restrictions = restrictions
+            self.output_keys[dts.key] = id
+
+    def unpack(self, id: ShuffleId, key: str) -> None:
+        "Handle an `unpack` task for a shuffle completing"
+        # Check if all output keys are done
+
+        # NOTE: we don't actually need this `unpack` step or tracking output keys;
+        # we could just delete the state in `barrier`.
+        # But we do it so we can detect duplicate shuffles, where a `transfer` task
+        # tries to reuse a shuffle ID that we're unpacking.
+        # (It does also allow us to clean up worker restrictions on error)
+        state = self.shuffles[id]
+        assert (
+            state.barrier_reached
+        ), f"Output {key} complete but barrier for shuffle {id} not yet reached"
+        assert (
+            state.out_tasks_left > 0
+        ), f"Output {key} complete; nout_left = {state.out_tasks_left} for shuffle {id}"
+
+        state.out_tasks_left -= 1
+
+        ts: TaskState = self.scheduler.tasks[key]
+        assert (
+            len(ts._worker_restrictions) == 1
+        ), f"Output {key} missing worker restrictions"
+        ts._worker_restrictions.clear()
+        del self.output_keys[key]
+
+        if not state.out_tasks_left:
+            # Shuffle is done. Yay!
+            del self.shuffles[id]
+
+    def erred(self, id: ShuffleId, key: str) -> None:
+        "Handle any task for a shuffle erroring"
+        try:
+            state = self.shuffles.pop(id)
+        except KeyError:
+            return
+
+        if state.barrier_reached:
+            # Remove worker restrictions for output tasks, in case the shuffle is re-submitted
+            for k, id_ in list(self.output_keys.items()):
+                if id_ == id:
+                    ts: TaskState = self.scheduler.tasks[k]
+                    ts._worker_restrictions.clear()
+                    del self.output_keys[k]
+
+    def transition(self, key: str, start: str, finish: str, *args, **kwargs):
+        "Watch transitions for keys we care about"
+        parts = parse_key(key)
```

Comment:
Annotations seem like the mechanism for this. Unfortunately there are two blockers to using them:
- https://github.com/dask/dask/issues/7036
- Blockwise fusion does not fuse across layers with different annotations. We need blockwise fusion to happen on the output tasks to prevent https://github.com/dask/distributed/issues/5223. Maybe we'd need some meta-annotation about whether an annotation can be fused?

So yes, it _should_ be easy, but it will require a bit of fixing and changing blockwise for it to actually be easy.

Reply:
I'm also thinking about getting rid of this transition-watching logic entirely, and having `transfer` tasks call an RPC on the scheduler to register themselves and get the list of peer workers (which would then also cause the scheduler to broadcast a message to all workers telling them the shuffle has started). This would eliminate the need for `parse_key` entirely. The only keys we'd have to parse are the output keys (for reasons mentioned in https://github.com/dask/distributed/pull/5524/files#r765203083). That would be more of a hybrid approach with https://github.com/dask/distributed/pull/5520.